### PR TITLE
feat: Add ATen lowering pass system

### LINKED ISSUE #2272: ✨[Feature] Upgrade Lowering Passes System Dynamo
- Add capabilities for subgraph rewrite passes both _before_ and _after_ ATen tracing
- Add new registries to hold implementations of both types of passes

Relevant to #1565, #2153

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Add documentation, testing, and lowering pass management systems for ATen lowering passes

Fixes #2272

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/passes/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+from .constant_folding import constant_fold
+from .repair_input_as_output import repair_input_as_output
+
+ATEN_LOWERING_PASSES = PassManager.build_from_passlist(
+    [
+        constant_fold,
+        repair_input_as_output,
+    ]
+)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/passes/__init__.py:11]: `add_lowering_pass` (function), `get_aten_lowering_passes` (function to get list of passes)

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/passes/__init__.py:11]: + Tutorial on how to build a lowering pass

File: py/torch_tensorrt/dynamo/lowering/passes/__init__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/passes/__init__.py:14]: `dump_current_lowering_passes` + `add_lowering_pass` takes an optional index arg to specify where in the list the lowering pass gets placed.

File: examples/dynamo/dynamo_aten_lowering_passes.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+This interactive script is intended as an overview of the process by which ATen lowering passes are written and used."""
+
+# %%
+# 1. Lowering Pass Function
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# An ATen lowering pass function in Torch-TRT must satisfy two requirements:
+# - The function must take as input a single `torch.fx.GraphModule` and return the lowered
+# `torch.fx.GraphModule`
```
Comment (narendasan) [inline examples/dynamo/dynamo_aten_lowering_passes.py:15]: Link to writing FX passes in pytorch docs

Comment (gs-olive) [inline examples/dynamo/dynamo_aten_lowering_passes.py:15]: Add a link to Torch FX passes instructions.

File: py/torch_tensorrt/dynamo/lowering/passes/__init__.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+) -> None:
+    """Adds a lowering pass to the registry, at a specified index if desired
+
+    If no index is specified, the lowering pass is inserted at the end of the list
+    """
+    ATEN_LOWERING_PASSES.add_pass_with_index(lowering_pass, index)
+    logger.debug(
+        f"Added lowering pass {lowering_pass} to list at index {index}, current passlist: {ATEN_LOWERING_PASSES}"
+    )
+    return
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/passes/__init__.py:33]: Make into a decorator

File: examples/dynamo/dynamo_aten_lowering_passes.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+# `torch_tensorrt.dynamo.lowering.passes`. See below for an example:
+
+# %%
+from torch_tensorrt.dynamo.lowering.passes import add_lowering_pass
+
+# Adds the lowering pass at the end of the pass list
+add_lowering_pass(repair_input_as_output)
+
+# Alternatively, specify an index to insert the lowering pass at a specific location
+add_lowering_pass(repair_input_as_output, 1)
```
Comment (gs-olive) [inline examples/dynamo/dynamo_aten_lowering_passes.py:93]: `index=1`

File: examples/dynamo/dynamo_aten_lowering_passes.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+# To remove a lowering pass, specify the index of the pass to remove:
+from torch_tensorrt.dynamo.lowering.passes import remove_lowering_pass
+
+# Removes the lowering pass at index 1
+remove_lowering_pass(1)
```
Comment (gs-olive) [inline examples/dynamo/dynamo_aten_lowering_passes.py:99]: `index=1`

File: py/torch_tensorrt/dynamo/lowering/passes/__init__.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def remove_lowering_pass(index: int) -> None:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/passes/__init__.py:36]: add `_` to denote internal-only

File: examples/dynamo/dynamo_aten_lowering_passes.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+
+# %%
+# 2. Lowering Pass Registration
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# To add a lowering pass, use the convenience function `add_lowering_pass` in the module
+# `torch_tensorrt.dynamo.lowering.passes`. See below for an example:
+
+# %%
```
Comment (gs-olive) [inline examples/dynamo/dynamo_aten_lowering_passes.py:86]: More about how to register in the pipeline if building Torch-TRT from source.

File: examples/dynamo/dynamo_aten_lowering_passes.py
Comment (gs-olive) [inline examples/dynamo/dynamo_aten_lowering_passes.py:1]: move to docsrc/contributors

File: py/torch_tensorrt/dynamo/lowering/passes/__init__.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+from typing import Callable, Optional, Tuple
+
+import torch
+
+# Import and order lowering passes and pass manager
+from .constant_folding import constant_fold
+from .pass_manager import DynamoPassManager
+from .repair_input_as_output import repair_input_as_output
+
+ATEN_LOWERING_PASSES = DynamoPassManager.build_from_passlist(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/lowering/passes/__init__.py:12]: Lets minimize code in __init__.py files, Lets create a `dynamo.lowering.passes._aten_lowering_pass` and import from there

File: py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    from torch._inductor.constant_folding import (
+        ConstantFolder,
+        replace_node_with_constant,
+    )
+
+logger = logging.getLogger(__name__)
+
+
+@torch.utils._python_dispatch._disable_current_modes()  # type: ignore
+def constant_fold(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py:21]: Thoughts on leveraging the same decorator API for our internal pipeline?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py:21]: I think the issue is it exposes the same issues we were talking about yesterday about registration errors

File: py/torch_tensorrt/dynamo/lowering/passes/pass_manager.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    def __init__(
+        self,
+        passes: Optional[
+            List[Callable[[torch.fx.GraphModule], torch.fx.GraphModule]]
+        ] = None,
+    ):
+        super().__init__(passes)
+
+    @classmethod
+    def build_from_passlist(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/lowering/passes/pass_manager.py:17]: n.b. for 2.2 we could build around this method for the pipeline creation API

File: docsrc/contributors/writing_dynamo_aten_lowering_passes.rst [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+    @aten_lowering_pass
+    def my_custom_pass(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
+        ...
+
+Alternatively, to insert the pass at a custom index (such as the front of the list) in the passlist, the following code can be used:
+
+.. code-block:: python
+
+    @aten_lowering_pass(index=0)
```
Comment (narendasan) [inline docsrc/contributors/writing_dynamo_aten_lowering_passes.rst:92]: lets add the _ in front (`_aten_lowering_pass`) to signify its an internal / unstable API

File: py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py (hunk 10 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py:21]: Discussed resolution

**Review by narendasan — APPROVED** [review]
LGTM, Can you create an issue to write an RFC for the api in 2.2?

**Comment by gs-olive** [comment]
Added in #2338
